File: libUtil/src/main/java/org/waynezhou/libUtil/LogHelper.java
package org.waynezhou.libUtil;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.waynezhou.libUtil.reflection.Reflection;
import org.waynezhou.libUtil.reflection.ReflectionException;

import java.util.Locale;

public final class LogHelper {
    private LogHelper() {
    }

    @NonNull
    private static String packageName = "org.waynezhou.libUtil";
    @NonNull
    private static String tag = "Log";

    static {
        try {
            packageName = Reflection
                    .getCurrentApplication()
                    .getApplicationInfo().packageName;
            String[] s = packageName.split("\\.");
            tag = s[s.length - 1] + tag;
        } catch (ReflectionException e) {
            Log.e(tag, "exception", e);
        }
    }

    @FunctionalInterface
    private interface LogMethod {
        void log(String tag, String msg);
    }

    @FunctionalInterface
    private interface LogThrowableMethod {
        void log(String tag, String msg, Throwable tr);
    }


    private static void log(LogMethod method, Object msg) {
        StackTraceElement callerInfo = Thread.currentThread().getStackTrace()[4];
        String objMsg = "null";
        if (msg != null) objMsg = msg.toString();
        objMsg = callerInfo.getClassName().replaceFirst(packageName, "")
                + "::"
                + callerInfo.getMethodName() + "(" + callerInfo.getFileName() + ":" + callerInfo.getLineNumber() + ")"
                + (objMsg.isEmpty()? "" : " - ")
                + objMsg;
        method.log(tag, objMsg);
    }

    private static void log(LogThrowableMethod method, Object msg, Throwable tr) {
        StackTraceElement callerInfo = Thread.currentThread().getStackTrace()[4];
        String objMsg = "null";
        if (msg != null) objMsg = msg.toString();
        objMsg = callerInfo.getClassName().replaceFirst(packageName, "")
                + "::"
                + callerInfo.getMethodName() + "(" + callerInfo.getFileName() + ":" + callerInfo.getLineNumber() + ")"
                + "\n"
                + objMsg;
        method.log(tag, objMsg, tr);
    }

    public static void ie(Condition isError, String format, Object... args){
        if(test(isError)){
            log(Log::e, String.format(Locale.TRADITIONAL_CHINESE, format, args));
        }else{
            log(Log::i, String.format(Locale.TRADITIONAL_CHINESE, format, args));
        }
    }
    public static void ie(Condition isError, @Nullable Object msg){
        if(test(isError)){
            log(Log::e, msg);
        }else{
            log(Log::i, msg);
        }
    }

    public static void d(String format, Object... args){
        log(Log::d, String.format(Locale.TRADITIONAL_CHINESE, format, args));
    }

    public static void d(@Nullable Object msg) {
        log(Log::d, msg);
    }

    @Deprecated
    public static void d(@Nullable Object msg, @NonNull Throwable tr) {
        log(Log::d, msg, tr);
    }

    public static void e(@Nullable Object msg) {
        log(Log::e, msg);
    }

    public static void e(@Nullable Object msg, @NonNull Throwable tr) {
        log(Log::e, msg, tr);
    }

    public static void i(){
        log(Log::i, "");
    }

    public static void i(String format, Object... args){
        log(Log::i, String.format(Locale.TRADITIONAL_CHINESE, format, args));
    }

    public static void i(@Nullable Object msg) {
        log(Log::i, msg);
    }

    @Deprecated
    public static void i(@Nullable Object msg, @NonNull Throwable tr) {
        log(Log::i, msg, tr);
    }

    public static void v(@Nullable Object msg) {
        log(Log::v, msg);
    }

    @Deprecated
    public static void v(@Nullable Object msg, @NonNull Throwable tr) {
        log(Log::v, msg, tr);
    }

    public static void w(@Nullable Object msg) {
        log(Log::w, msg);
    }

    @Deprecated
    public static void w(@Nullable Object msg, @NonNull Throwable tr) {
        log(Log::w, msg, tr);
    }

    public static void wtf(@Nullable Object msg) {
        log(Log::wtf, msg);
    }

    @Deprecated
    public static void wtf(@Nullable Object msg, @NonNull Throwable tr) {
        log(Log::wtf, msg, tr);
    }

    private static boolean test(Condition condition){
        return condition.test();
    }
    @FunctionalInterface
    public interface Condition{
        boolean test();
    }
}

File: libUtilKt/src/main/java/org/waynezhou/libutilkt/LogHelper.kt
package org.waynezhou.libutilkt

import android.util.Log
import java.util.*


typealias LogMethod = (tag: String?, msg: String) -> Unit
typealias LogThrowableMethod = (tag: String?, msg: String?, tr: Throwable?) -> Unit

@Suppress("unused")
object LogHelper {
    private fun log(log: LogMethod, msg: Any?) {
        val callerInfo = Thread.currentThread().stackTrace[4]
        var objMsg = msg?.toString() ?: "null"
        objMsg =
            (callerInfo.methodName + "(" + callerInfo.fileName + ":" + callerInfo.lineNumber + ")"
                    + (if (objMsg.isEmpty()) "" else " - ")
                    + objMsg)
        log("LogHelper", objMsg)
    }

    private fun log(log: LogThrowableMethod, msg: Any?, tr: Throwable) {
        val callerInfo = Thread.currentThread().stackTrace[4]
        var objMsg = msg?.toString() ?: "null"
        objMsg =
            (callerInfo.methodName + "(" + callerInfo.fileName + ":" + callerInfo.lineNumber + ")"
                    + (if (objMsg.isEmpty()) "" else " - ")
                    + objMsg)
        log("LogHelper", objMsg, tr)
    }

    fun ie(isError: () -> Boolean, format: String, vararg args: Any?) {
        if (isError())
            log(Log::e, String.format(Locale.TRADITIONAL_CHINESE, format, *args))
        else
            log(Log::i, String.format(Locale.TRADITIONAL_CHINESE, format, *args))
    }

    fun ie(isError: () -> Boolean, msg: Any?) {
        if (isError())
            log(Log::e, msg)
        else
            log(Log::i, msg)
    }

    fun d(format: String, vararg args: Any?) {
        log(Log::d, String.format(Locale.TRADITIONAL_CHINESE, format, *args))
    }

    fun d(msg: Any?) {
        log(Log::d, msg)
    }

    fun d(tr: Throwable, msg: Any?) {
        log(Log::d, msg, tr)
    }

    fun e(msg: Any?) {
        log(Log::e, msg)
    }

    fun e(format: String, vararg args: Any?) {
        log(Log::e, String.format(Locale.TRADITIONAL_CHINESE, format, *args))
    }

    fun e(tr: Throwable, msg: Any?) {
        log(Log::e, msg, tr)
    }

    fun i() {
        log(Log::i, "")
    }

    fun i(format: String, vararg args: Any?) {
        log(Log::i, String.format(Locale.TRADITIONAL_CHINESE, format, *args))
    }

    fun i(msg: Any?) {
        log(Log::i, msg)
    }

    fun i(tr: Throwable, msg: Any?) {
        log(Log::i, msg, tr)
    }

    fun v(msg: Any?) {
        log(Log::v,  msg)
    }

    fun v(tr: Throwable, msg: Any?) {
        log(Log::v, msg, tr)
    }

    fun w(msg: Any?) {
        log(Log::w, msg)
    }

    fun w(format: String, vararg args: Any?) {
        log(Log::w, String.format(Locale.TRADITIONAL_CHINESE, format, *args))
    }

    fun w(tr: Throwable, msg: Any?) {
        log(Log::w, msg, tr)
    }

    fun wtf(msg: Any?) {
        log(Log::wtf, msg)
    }

    fun wtf(tr: Throwable, msg: Any?) {
        log(Log::wtf, msg, tr)
    }

    /*@NonNull
    private static String tag = "Log";

    static {
        try {
            String packageName = Reflection
              .getCurrentApplication()
              .getApplicationInfo().packageName;
            String[] s = packageName.split("\\.");
            tag = s[s.length - 1] + tag;
        } catch (ReflectionException e) {
            Log.e(tag, "exception", e);
        }
    }*/

}
